core/src/main/java/com/linecorp/armeria/common/DefaultHttpResponseBodyStream.java

```diff
@@ -0,0 +1,352 @@
+    }
+
+    @Override
+    public CompletableFuture<HttpHeaders> trailers() {
+        final HeadersFuture<HttpHeaders> trailersFuture = this.trailersFuture;
+        if (trailersFuture != null) {
+            return trailersFuture;
+        }
+
+        trailersFutureUpdater.compareAndSet(this, null, new HeadersFuture<>());
+        return this.trailersFuture;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return response.isOpen();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return !isOpen() && !wroteAny;
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return response.whenComplete();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor unused) {
+        requireNonNull(subscriber, "subscriber");
+        if (executor.inEventLoop()) {
+            bodySubscriber.setDownStream(subscriber);
+        } else {
+            executor.execute(() -> bodySubscriber.setDownStream(subscriber));
+        }
+    }
+
+    @Override
+    public void abort() {
+        response.abort();
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        response.abort(cause);
+    }
+
+    private final class BodySubscriber implements Subscriber<HttpObject>, Subscription {
+
+        @Nullable
+        private ImmutableList.Builder<ResponseHeaders> informationalHeadersBuilder;
+        @Nullable
+        private Throwable cause;
+        private boolean completing;
+
+        private boolean sawLeadingHeaders;
+
+        @Nullable
+        volatile Subscriber<? super HttpData> downstream;
+        @Nullable
+        volatile Subscription upstream;
+        private volatile long pendingRequests;
+        private volatile boolean cancelCalled;
+
+        private void setDownStream(Subscriber<? super HttpData> downstream) {
+            try {
+                if (!downstreamUpdater.compareAndSet(this, null, downstream)) {
+                    downstream.onSubscribe(NoopSubscription.get());
+                    downstream.onError(new IllegalStateException("subscribed by other subscriber already"));
+                    return;
+                }
+                downstream.onSubscribe(this);
+                if (cause != null) {
+                    downstream.onError(cause);
+                } else if (completing) {
+                    downstream.onComplete();
+                }
+            } catch (Throwable t) {
+                throwIfFatal(t);
+                logger.warn("Subscriber should not throw an exception. subscriber: {}", downstream, t);
+            }
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (!upstreamUpdater.compareAndSet(this, null, subscription) || cancelCalled) {
+                subscription.cancel();
+                return;
+            }
+            if (pendingRequests != 0) {
```

Comment:
I think there's a race with the `request` method.
For example:
- A thread is in the line between 215 and 216
- B thread calls `subscription.request(pendingRequests);`
- A thread does `pendingRequests = LongMath.saturatedAdd(n, pendingRequests);`

Reply:
How about just using the single thread to avoid this problem?

Reply:
> How about just using the single thread to avoid this problem?

That's sounds good. It is no additional threading because the `upstream.request()` already is executed `subscription.executor()`.